crates/services/executor/src/executor.rs

```diff
@@ -624,7 +659,144 @@ where
+        let block_height = *block.header.height();
+        let execution_kind = ExecutionKind::Validation;
+
+        let forced_transactions = if self.relayer.enabled() {
+            self.process_da(&block.header, &mut data)?
+        } else {
+            Vec::with_capacity(0)
+        };
+
+        // The block level storage transaction that also contains data from the relayer.
+        // Starting from this point, modifications from each thread should be independent
+        // and shouldn't touch the same data.
+        let mut block_with_relayer_data_transaction = self.block_st_transaction.read_transaction()
+            // Enforces independent changes from each thread.
+            .with_policy(ConflictPolicy::Fail);
+
+        // We execute transactions in a single thread right now, but later,
+        // we will execute them in parallel with a separate independent storage transaction per thread.
+        let mut thread_block_transaction = block_with_relayer_data_transaction
+            .read_transaction()
+            .with_policy(ConflictPolicy::Overwrite);
+
+        debug_assert!(block.transactions.is_empty());
+
+        let relayed_tx_iter = forced_transactions.into_iter();
+        for transaction in relayed_tx_iter {
+            const RELAYED_GAS_PRICE: Word = 0;
+            let transaction = MaybeCheckedTransaction::CheckedTransaction(transaction);
+            let tx_id = transaction.id(&self.consensus_params.chain_id());
+            match self.execute_transaction_and_commit(
+                block,
+                &mut thread_block_transaction,
+                &mut data,
+                transaction,
+                RELAYED_GAS_PRICE,
+                coinbase_contract_id,
+                execution_kind,
+            ) {
+                Ok(_) => {}
+                Err(err) => {
+                    let event = ExecutorEvent::ForcedTransactionFailed {
+                        id: tx_id.into(),
+                        block_height,
+                        failure: err.to_string(),
+                    };
+                    data.events.push(event);
+                }
+            }
+        }
+
+        // L2 originated transactions should be in the `TxSource`. This will be triggered after
+        // all relayed transactions are processed.
+        let mut regular_tx_iter = source.next(block_gas_limit).into_iter().peekable();
+        while regular_tx_iter.peek().is_some() {
+            for transaction in regular_tx_iter {
+                self.execute_transaction_and_commit(
+                    block,
+                    &mut thread_block_transaction,
+                    &mut data,
+                    transaction,
+                    gas_price,
+                    coinbase_contract_id,
+                    execution_kind,
+                )?;
+            }
+
+            let new_remaining_gas_limit = block_gas_limit.saturating_sub(data.used_gas);
```

Comment:
Hmm, instead of `saturating_sub` it should be `checked_add`. Because maybe bloc producer included more transactions than was allowed=)

Also, maybe we need to update `execute_inner` as well

Reply:
In the follow-up PR, I don't event have this variable:
https://github.com/FuelLabs/fuel-core/pull/1847

I just expect the execution to fail if it exceeds the limit.